NE-1951: Pre-upgrade Admin Gate for Gateway API CRD Management Succession

Thealisyed opened the pull request:
This PR introduces the following:

Initialize the GatewayAPI upgradeable controller
- Added the required imports and constants.
- Defined predicates for CRD and ConfigMap events to manage the admin gate.
- Configured watches in the controller for specific changes in the ConfigMap.

Implement the reconciler logic
- Added the Reconcile function to manage the admin gate addition and removal based on conditions

Enhance the operator initialization
- Conditionally initialize the gatewayapi_upgradeable controller based on the GatewayAPI feature gate.

 End-to-End Tests
- Added the TestGatewayAPIUpgradeable end-to-end test.
- Included the new test case in the test suite.

Thealisyed commented on pkg/operator/controller/gatewayapi/controller.go:
```diff
@@ -83,6 +83,9 @@ type Config struct {
 	DependentControllers []controller.Controller
+
+	//AdminConsentRequired indicates
+	AdminConsentRequired bool
```
Not needed, will delete!

Thealisyed commented:
Used Grant's [PR](https://github.com/openshift/cluster-ingress-operator/pull/1014/commits/6778ce1a619b2f58310a037694e341e446888c55#) as reference to implement admin gate in this manner.

Miciah commented:
/assign
/assign @shaneutt

Thealisyed commented on pkg/operator/controller/names.go:
```diff
@@ -49,6 +49,13 @@ const (
+func AdminGatesConfigMapName() types.NamespacedName {
+	return types.NamespacedName{
+		Name:      "admin-gates",
+		Namespace: "openshift-config",
```
use GlobalMachineSpecifiedConfigNamespace instead. 
OP Config managed is what the operator created

Thealisyed commented on pkg/operator/operator.go:
```diff
@@ -3,6 +3,7 @@ package operator
 import (
 	"context"
 	"fmt"
+	gatewayapi_upgradeable "github.com/openshift/cluster-ingress-operator/pkg/operator/controller/gatewayapi-upgradeable"
```
group these properly, should be in the subsequent box

Thealisyed commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,183 @@
+const (
+	controllerName     = "gatewayapi_upgradeable_controller"
+	GatewayAPIAdminKey = "ack-gateway-api-management"
+	GatewayAPIAdminMsg = "You are now acknowledging that you are taking control of the Gateway API CRD Management from here on out. Please review the documentation for implications."
```
Reference the correct documentation and fix the message

Thealisyed commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,183 @@
+
+	// Watch for CRD updates using the reconciler cache and apply the CRD predicate
+	if err := c.Watch(source.Kind(reconciler.cache, &apiextensionsv1.CustomResourceDefinition{}),
+		&handler.EnqueueRequestForObject{},
```
the kinds should be swapped for the watches you have.

alebedev87 replied:
What does your controller reconcile? ConfigMaps or CRDs? IMO, this watch should do the mapping to the admingate configmap. This way once you have GWAPI CRDs the admingate configmap is re-enrsured.

Thealisyed replied:
Amended this in the latest commit

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,183 @@
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	operatorCache := mgr.GetCache()
+	reconciler := &reconciler{
+		cache:  config.Cache,
```
The cache you passed in the operator is the manager's cache. No need to spawn many cache variables or fields, if `mgr.GetCache()` is what you need then `mgr` parameter is enough.

Thealisyed replied:
Fixed

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	// Define a predicate filter to watch for specific changes in the ConfigMap
+	adminGateFilter := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			// Watch for creation of the "admin-gates" ConfigMap in "openshift-config-managed" namespace
+			configmap := e.Object.(*corev1.ConfigMap)
+			return configmap.Name == operatorcontroller.AdminGatesConfigMapName().Name && configmap.Namespace == operatorcontroller.AdminGatesConfigMapName().Namespace
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool { return false }, // Ignore delete events
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// Watch for updates to the "admin-gates" ConfigMap to check if the admin key changes
+			configmapOld := e.ObjectOld.(*corev1.ConfigMap)
+			configmapNew := e.ObjectNew.(*corev1.ConfigMap)
+			if configmapOld.Name != operatorcontroller.AdminGatesConfigMapName().Name || configmapOld.Namespace != operatorcontroller.AdminGatesConfigMapName().Namespace {
+				return false
+			}
+			oldVal, oldExists := configmapOld.Data[GatewayAPIAdminKey]
+			newVal, newExists := configmapNew.Data[GatewayAPIAdminKey]
+			return oldExists != newExists || oldVal != newVal
+		},
+		GenericFunc: func(genericEvent event.GenericEvent) bool { return false }, // Ignore generic events
+	}
```
I think something like this should be enough:
```suggestion
	adminGatePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
	         cm := o.(*corev1.ConfigMap)
		return cm.Name ==  operatorcontroller.AdminGatesConfigMapName().Name && cm.Namespace == operatorcontroller.AdminGatesConfigMapName().Namespace
	})
```

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,183 @@
+
+	//TODO: too many args to c.watch and not enough to source.Kind
+	if err := c.Watch(source.Kind(operatorCache, &corev1.ConfigMap{}),
+		handler.EnqueueRequestsFromMapFunc(toAdminGateConfigMap),
```
Why it's not just `&handler.EnqueueRequestForObject{}`?  I think your controller should reconcile the admin gate configmap.

Thealisyed replied:
Fixed

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,183 @@
+
+// Config struct holds the cache information
+type Config struct {
+	//namespace referes to namespace where admin gates controller resides
+	Namespace string
```
I don't think you need this field (and maybe the whole config?) because the admingate configmap is already defined as a constant in `names.go`.

Thealisyed replied:
Correct, this is redundant

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,183 @@
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to determine if UnservableInFutureVersionsRoute routes exist: %w", err)
+	}
+	if adminGateConditionExists {
+		if err := r.addAdminGate(ctx); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to add admin gate: %w", err)
+		}
+	} else {
+		if err := r.removeAdminGate(ctx); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to remove admin gate: %w", err)
+		}
+	}
```
I suppose that the decision about adding or removing a data field to/from the admin in gate should be based on the presence of GWAPI CRDs which I don't see here.

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
The name of the controller doesn't seem to quite reflect what the controller is supposed to do. The controller will ensure the admingate for the gatewayAPI CRDs.

alebedev87 replied:
Sorry, I think I see where the inspiration is coming from: `route_upgradable` in `release-4.15` branch. Ok then, it may make sense to keep this kinda consistent.

alebedev87 reviewed (commented):
Good start! Keep it up! Some comments to steer the direction of the controller's implementation.

Thealisyed commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+// adminGateConditionExists checks if the admin gate condition exists based on both ConfigMap and CRDs
+func (r *reconciler) adminGateConditionExists(ctx context.Context) (bool, error) {
```
Function should check if the CRD's exist or not only!

Thealisyed commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,174 @@
+
+// adminGateConditionExists checks if the admin gate condition exists based on both ConfigMap and CRDs
+func (r *reconciler) adminGateConditionExists(ctx context.Context) (bool, error) {
+	adminGateConfigMap := &corev1.ConfigMap{}
+	if err := r.cache.Get(ctx, operatorcontroller.AdminGatesConfigMapName(), adminGateConfigMap); err != nil {
+		return false, fmt.Errorf("failed to get configmap %s: %w", operatorcontroller.AdminGatesConfigMapName(), err)
+	}
+
+	// Check if the admin key exists and is set to the expected message
+	if val, ok := adminGateConfigMap.Data[GatewayAPIAdminKey]; ok && val == GatewayAPIAdminMsg {
+		return true, nil // Exists as expected
+	}
```
Copy and paste as AddAdminGate?...

candita commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,164 @@
+
+import (
+	"context"
+	"fmt"
```
Add lines between package groups imports e.g. between standard packages, openshift packages, and k8s packages.

Miciah replied:
Still missing the blank line here.

Thealisyed replied:
@shaneutt Thanks for sharing this, will look into it!

candita commented on pkg/operator/controller/names.go:
```diff
@@ -49,6 +49,13 @@ const (
+func AdminGatesConfigMapName() types.NamespacedName {
+	return types.NamespacedName{
+		Name:      "admin-gates",
+		Namespace: GlobalMachineSpecifiedConfigNamespace,
```
Where is GlobalMachineSpecifiedConfigNamespace declared?

candita commented on pkg/operator/controller/names.go:
```diff
@@ -49,6 +49,13 @@ const (
 
+func AdminGatesConfigMapName() types.NamespacedName {
+	return types.NamespacedName{
+		Name:      "admin-gates",
```
We probably want to call this something like:  `openshift-ingress-admingates`

alebedev87 replied:
I don't think the name of the admin gate configmap can be customizable. [The EP](https://github.com/openshift/enhancements/blob/master/enhancements/update/upgrades-blocking-on-ack.md) mentions only `admin-gates` configmap.

candita commented on pkg/operator/operator.go:
```diff
@@ -319,6 +320,11 @@ func New(config operatorconfig.Config, kubeConfig *rest.Config) (*Operator, erro
 
+	// Set up the gateway api upgradeable controller
+	if _, err := gatewayapi_upgradeable.New(mgr); err != nil {
+		return nil, fmt.Errorf("failed to create gatewayapi upgradeable controller: %w", err)
```
```suggestion
		return nil, fmt.Errorf("failed to create gatewayapi upgradeable controller: %v", err)
```

candita commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+const (
+	controllerName     = "gatewayapi_upgradeable_controller"
+	GatewayAPIAdminKey = "ack-gateway-api-management"
+	GatewayAPIAdminMsg = "You are now acknowledging that you are taking control of the Gateway API CRD Management from here on out. Please review the documentation for implications."
```
The message should include three pieces of info: why the message was generated, what the issue is, and how seriously we advise against what they're trying to do.

candita commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New function initializes the controller and sets up the watch for the ConfigMap
```
It's a nit, but all comments should be complete sentences and end in a period.

candita commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+	// Check if the admin gate should be added or removed
+	adminGateConditionExists, err := r.adminGateConditionExists(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to determine if admin gate exist: %w", err)
```
```suggestion
		return reconcile.Result{}, fmt.Errorf("failed to determine if admin gate condition exists: %v", err)
```

candita commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+
+	for _, crd := range crds.Items {
+		if crd.Spec.Group == gatewayapiv1.GroupName {
+			return true, nil
```
Should we check the versions? If it is the same version, that should be less of a problem than if it were an earlier or later version, right?  It would impact the kind of error message we write.

Thealisyed replied:
According to the [EP](https://github.com/openshift/enhancements/pull/1756/files?short_path=e820058#diff-e82005837f3b20b3e0c690dfbe6dacac0581f41812783e633b73b1ffb030f16b:~:text=IF%20the%20CRDs,of%20CRD%20succession), the admingate is created upon the presense of the GWAPI CRD's (regardless of the versions).

candita commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+func (r *reconciler) addAdminGate(ctx context.Context) error {
+	adminGateConfigMap := &corev1.ConfigMap{}
+	if err := r.client.Get(ctx, operatorcontroller.AdminGatesConfigMapName(), adminGateConfigMap); err != nil {
+		return fmt.Errorf("failed to get configmap %s: %w", operatorcontroller.AdminGatesConfigMapName(), err)
```
What if it doesn't exist and we need to create it?  Can that happen?

Thealisyed replied:
The CVO is responsible for recreating  and managing the [ConfigMap](https://github.com/openshift/enhancements/blob/master/enhancements/update/upgrades-blocking-on-ack.md#:~:text=Another%20risk%20is,has%20been%20removed.) 
So if I understand correctly the if the ConfigMap is bogus or missing etc, the CVO will reconcile and fix it accordingly
Also in favor of following the same logic to implement this as done in this [PR](https://github.com/openshift/cluster-ingress-operator/pull/1014/commits/6778ce1a619b2f58310a037694e341e446888c55).

Miciah commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+func (r *reconciler) adminGateConditionExists(ctx context.Context) (bool, error) {
+	// Check for the presence of Gateway API CRDs
+	crds := &apiextensionsv1.CustomResourceDefinitionList{}
+	if err := r.client.List(ctx, crds); err != nil {
```
Use `r.cache` if you can.  Using `r.Client` puts more load on the API.

alebedev87 replied:
Just a note of clarification for @Thealisyed. By default the controller-runtime's client is a split client meaning that it makes read-only requests (get/list) against the downstream cache and write requests directly against the api server. However the cluster ingress operator [customizes the client](https://github.com/openshift/cluster-ingress-operator/blob/e80fa46b72ec08e6d8a7fc7f111d25842cbb80bf/pkg/operator/operator.go#L154-L165) to make all requests against the api server.

Thealisyed replied:
Fixed to use `r.cache` now

Thealisyed commented:
/remove-hold

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+func TestGatewayAPIUpgradeable(t *testing.T) {
+	t.Parallel()
+
+	t.Log("Starting test: Ensure the Gateway API CRDs exist")
+	ensureCRDs(t)
```
Your PR targets 4.18 branch which doesn't have `GatewayAPI` enabled. So, this part will fail on most of e2e tests (except for `e2e-operator-techpreview`).

alebedev87 replied:
As we discussed, you need to create a new function which creates a test CRD from gateway group (`.spec.group`).

Thealisyed replied:
Latest commit addresses this

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+
+	// Defer the cleanup of the test gateway.
+	t.Cleanup(func() {
+		t.Log("Starting cleanup: Deleting test gateway")
```
Deleting test gateway? You didn't create any gateway. You should check the code you copy from other tests whether it suits your need.

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+	})
+}
+
+func getClient() (*kubernetes.Clientset, error) {
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(kubeConfig)
+}
```
You don't need to create another instance of the client. There is [a global variable](https://github.com/openshift/cluster-ingress-operator/blob/4db86485a0c44e8592c993a8215b5d8ad6b44947/test/e2e/operator_test.go#L120) for this.

Thealisyed replied:
Fixed this in latest commit

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+func TestGatewayAPIUpgradeable(t *testing.T) {
+	t.Parallel()
+
+	t.Log("Starting test: Ensure the Gateway API CRDs exist")
```
In CIO repository, the tests tend to be less verbose. We try to limit logs to things which fail (e.g. retries) or important to the logic of the test (featuregates, parameter changes, etc.). You can keep those logs are comments though.

Thealisyed replied:
Fixed

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+	t.Log("Ensuring CRDs exist")
+	crdsExist := true
+
+	t.Logf("Getting configmap admin-gates in namespace openshift-config-managed")
+	configMap, err := kclient.CoreV1().ConfigMaps("openshift-config-managed").
+		Get(context.TODO(), "admin-gates", metav1.GetOptions{})
+	if err != nil {
+		t.Errorf("failed to get configmap admin-gates: %v", err)
+		if errors.IsNotFound(err) {
+			t.Logf("ConfigMap 'admin-gates' not found in namespace openshift-config-managed")
+		} else {
+			t.Logf("Error getting ConfigMap: %v", err)
+		}
+		return
+	}
```
Can you please wrap this into a polling loop? See other tests, how they make use of `wait.PollUntilContextTimeout` function.

Thealisyed replied:
Makes sense and I have implemented this in the latest commit

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+		t.Fatalf("failed to get kube client: %v", err)
+	}
+
+	t.Log("Ensuring CRDs exist")
+	crdsExist := true
```
How can this impact the check of the admingate?

Thealisyed replied:
Right, this was not correct as it was hardcoding the value, was fixed in the latest commit

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+		adminGateKeyExists = true
+	}
+
+	t.Logf("CRDs exist: %v, Admin Gate Key exists: %v", crdsExist, adminGateKeyExists)
```
This is ok to log while you debug your test but, it's too verbose for a job which will be run as a pre-submit for every PR.

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+
+	t.Logf("CRDs exist: %v, Admin Gate Key exists: %v", crdsExist, adminGateKeyExists)
+
+	if crdsExist && !adminGateKeyExists {
```
Again, I don't quite get how `crdExist` influences the decision about the admingate.

Thealisyed replied:
Right this was not implemented properly in this commit however overall if gwapi crd are present operator will add admin gate

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,94 @@
+	} else if !crdsExist && adminGateKeyExists {
+		t.Errorf("expected admin gate key to be removed from configmap, but it still exists")
+	}
+}
```
You use the same `testAdminGate` function to validate the existence of the expected admin gate for its absence (after the crd deletion). How is it supposed to work?

Thealisyed replied:
The logic for this is not accurate, have updated and fixed this in the latest commit

alebedev87 reviewed (changes requested):
I had a look at e2e test. Good start but it needs some modifications.

shaneutt commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,166 @@
+		return nil, err
+	}
+
+	if err := c.Watch(source.Kind[client.Object](mgr.GetCache(), &corev1.ConfigMap{}, &handler.EnqueueRequestForObject{}, adminGatePredicate)); err != nil {
```
nit: you could move the definition of the `adminGatePredicate` right above here since this is the only place its used.

```suggestion
	adminGatePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
		return o.GetNamespace() == operatorcontroller.AdminGatesConfigMapName().Namespace &&
			o.GetName() == operatorcontroller.AdminGatesConfigMapName().Name
	})

	if err := c.Watch(source.Kind[client.Object](mgr.GetCache(), &corev1.ConfigMap{}, &handler.EnqueueRequestForObject{}, adminGatePredicate)); err != nil {
```

shaneutt reviewed (commented):
I took a very brief initial first pass and have some small suggestions. I intend to do another pass that's thorough in the morning.

Important note: whenever you see a "nit" comment from me, please consider it resolvable at your discretion. These are not items I feel need to hold anything up.

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,124 @@
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+func kubeinit() {
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		panic(err)
+	}
+	kclient, err = client.New(kubeConfig, client.Options{})
+	if err != nil {
+		panic(err)
+	}
+}
```
Please use [the package global kclient variable](https://github.com/openshift/cluster-ingress-operator/blob/f1b159ad1f30f2a3034e13b264081f08e67695b4/test/e2e/operator_test.go#L120) instead.

Thealisyed replied:
So initially, I used kclient (for whatever reason), however when I looked around other test files, I found the standard practice for retrieving the kubeconfig as follows:

```
kubeConfig, err := config.GetConfig()
	if err != nil {
		t.Fatalf("failed to get kube config: %v", err)
	}
	client, err := kubernetes.NewForConfig(kubeConfig)
	if err != nil {
		t.Fatalf("failed to create kube client: %v", err)
	}
```

To maintain consistency with the existing tests, I have adopted this approach in my code.
Does this approach work, or should I revert to using the package global kclient variable instead?

alebedev87 replied:
Just remove `kubeinit()` completely, you don't need it.

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,124 @@
+func TestGatewayAPIUpgradeable(t *testing.T) {
+	t.Parallel()
+
+	//t.Log("Starting test: Ensure the Gateway API CRDs exist")
```
What I meant was:

```suggestion
	// Starting test: Ensure the Gateway API CRDs exist
```

Thealisyed replied:
I removed the excessive logs and comments, just kept what is needed for simplicity

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,124 @@
+	configMap := &corev1.ConfigMap{}
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 30*time.Second, false, func(ctx context.Context) (bool, error) {
+		if err := kclient.Get(ctx, client.ObjectKey{Namespace: "openshift-config-managed", Name: "admin-gates"}, configMap); err != nil {
+			//t.Logf("Failed to get configmap admin-gates: %v", err)
+			if errors.IsNotFound(err) {
+				//t.Logf("ConfigMap 'admin-gates' not found in namespace openshift-config-managed")
+			} else {
+				//t.Logf("Error getting ConfigMap: %v", err)
+			}
```
This log message, on the other hand, does make sense as it signals a problem:
```suggestion
			t.Logf("Failed to get configmap admin-gates: %v, retrying...", err)
```
And yes, you don't need to make a distinction between error types because the error will contain the needed description. E.g. if it's "NotFound" error, the description will say so.

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,124 @@
+		return true, nil
+	})
+	if err != nil {
+		t.Errorf("Failed to get configmap admin-gates after polling: %v", err)
```
```suggestion
		t.Errorf("Timed out trying to get admin-gates configmap: %v", err)
```

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,124 @@
+		return
+	}
+
+	adminGateKeyExists := false
+	if _, ok := configMap.Data[gatewayapi_upgradeable.GatewayAPIAdminKey]; ok {
+		adminGateKeyExists = true
+		//t.Log("Admin gate key exists in the configmap")
+	} else {
+		//	t.Log("Admin gate key does not exist in the configmap")
+	}
+	if crdsExist != shouldExist {
+		t.Errorf("Expected CRDs existence: %v, but got: %v", shouldExist, crdsExist)
+	}
+	if adminGateKeyExists != shouldExist {
+		t.Errorf("Expected admin gate key existence: %v, but got: %v", shouldExist, adminGateKeyExists)
+	}
```
I still don't get why you check for the existence of CRDs, you ensured their existence as the first step of `TestGatewayAPIUpgradeable` test (`createCRDs` function).
```suggestion
	_, adminGateKeyExists := configMap.Data[gatewayapi_upgradeable.GatewayAPIAdminKey]
        return adminGateKeyExists
```

Thealisyed replied:
Fixed

alebedev87 reviewed (commented):
Another look at the e2e.

shaneutt commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,124 @@
+//go:build e2e
```
Throughout this file there are several commented out log lines that would appear to need to be uncommented, or changed into regular comments?

Thealisyed replied:
Correct, I removed the excessive logs and comments, keeping only what is needed

shaneutt commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,166 @@
+
+	// Defining the CRD predicate.
+	crdPredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.(*apiextensionsv1.CustomResourceDefinition).Spec.Group == gatewayapiv1.GroupName
```
It occurs to me that we technically need to cover the new [gateway.networking.x-k8s.io](https://github.com/kubernetes-sigs/gateway-api/blob/2b0b9ceff08f1ac7899efa25657480d3196b1185/config/crd/experimental/gateway.networking.x-k8s.io_listenersets.yaml#L11) group here as well.

```suggestion
		group := o.(*apiextensionsv1.CustomResourceDefinition).Spec.Group
		return group == gatewayapiv1.GroupName || group == "gateway.networking.x-k8s.io"
```

(note: feel free to merge this suggestion, but don't use the suggestion verbatim switch to an import)

Thealisyed replied:
Fixed

shaneutt reviewed (commented):
Second pass!

Thealisyed commented:
/retest 

TestNodePool and unrelated to changes

Thealisyed commented:
/retest 
Node installer timeout

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+		t.Skip("Gateway API is enabled, skipping TestGatewayAPIUpgradeable")
+	}
+
+	t.Parallel()
```
This is better to be at the very start of the test function.

Thealisyed replied:
Done

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+		if err := kclient.Create(context.TODO(), crd); err != nil {
+			if !errors.IsAlreadyExists(err) {
+				t.Fatalf("Failed to create CRD %s: %v", crd.Name, err)
+			}
+			continue
```
```suggestion
			}
```

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+func testAdminGate(t *testing.T, shouldExist bool) {
+	t.Helper()
+	//	crdsExist := checkCRDsExist(t)
+	configMap := &corev1.ConfigMap{}
```
```suggestion
	adminGatesConfigMap := &corev1.ConfigMap{}
```

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+	t.Helper()
+	//	crdsExist := checkCRDsExist(t)
+	configMap := &corev1.ConfigMap{}
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 30*time.Second, false, func(ctx context.Context) (bool, error) {
+		if err := kclient.Get(ctx, client.ObjectKey{Namespace: "openshift-config-managed", Name: "admin-gates"}, configMap); err != nil {
+			t.Logf("Failed to get configmap admin-gates: %v, retrying...", err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Errorf("Timed out trying to get admin-gates configmap: %v", err)
+		return
+	}
```
```suggestion
	if err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 30*time.Second, false, func(ctx context.Context) (bool, error) {
		if err := kclient.Get(ctx, client.ObjectKey{Namespace: "openshift-config-managed", Name: "admin-gates"}, configMap); err != nil {
			t.Logf("Failed to get configmap admin-gates: %v, retrying...", err)
			return false, nil
		}
		return true, nil
	}); err != nil {
		t.Fatalf("Timed out trying to get admin-gates configmap: %v", err)
	}
```

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+
+	_, adminGateKeyExists := configMap.Data[gatewayapi_upgradeable.GatewayAPIAdminKey]
+	if adminGateKeyExists != shouldExist {
+		t.Errorf("Expected admin gate key existence to be %v, but got %v", shouldExist, adminGateKeyExists)
```
```suggestion
		t.Fatalf("Expected admin gate key existence to be %v, but got %v", shouldExist, adminGateKeyExists)
```

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+	log = logf.Logger.WithName(controllerName)
+)
+
+var kclient client.Client
```
What is it for?! Didn't use mix this with e2e test accidentally? Your controller will use the client from the controller-runtime's manager.

Thealisyed replied:
Apologies, seems to be a mix up, fixed now!

alebedev87 commented on pkg/operator/operator.go:
```diff
@@ -319,6 +319,13 @@ func New(config operatorconfig.Config, kubeConfig *rest.Config) (*Operator, erro
 		return nil, fmt.Errorf("failed to create gatewayapi controller: %w", err)
 	}
 
+	// Add conditional setup for gateway_upgradeable controller only if FeatureGate is enabled
```
The comment is telling the opposite of what is happening.

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+const (
+	controllerName     = "gatewayapi_upgradeable_controller"
+	GatewayAPIAdminKey = "ack-gateway-api-management"
+	GatewayAPIAdminMsg = "Gateway API CRDs have been detected. OCP fully manages the life-cycle of Gateway API CRDs. External management is unsupported and will be prevented. Tampering with the prevention mechanism will cause the cluster to move to a 'Degraded' state. The cluster administrator is responsible for the safety of existing Gateway API implementations and must acknowledge their responsibi ...
```
We agreed on the doc link during our last call: `https://docs.openshift.com/container-platform/4.19/release_notes/ocp-4-19-release-notes.html#ocp-4-19-networking-gateway-api-crd-lifecycle_release-notes`. Can you please confirm @grzpiotrowski?

grzpiotrowski replied:
Since `docs.openshift.com` is decommissioned in favour of `docs.redhat.com` the doc link should most likely be:
`https://docs.redhat.com/en/documentation/openshift_container_platform/4.19/release_notes/ocp-4-19-release-notes.html#ocp-4-19-networking-gateway-api-crd-lifecycle_release-notes`

shaneutt replied:
```suggestion
	GatewayAPIAdminMsg = "Gateway API CRDs have been detected. OCP fully manages the life-cycle of Gateway API CRDs. External management is unsupported and will be prevented. Tampering with the prevention mechanism will cause the cluster to move to a 'Degraded' state. The cluster administrator is responsible for the safety of existing Gateway API implementations and must acknowledge their responsibilities via the admin gate to proceed with upgrades. See https://docs.redhat.com/en/documentation/openshift_container_platform/4.19/html/release_notes/ocp-4-19-release-notes#ocp-4-19-networking-gateway-api-crd-lifecycle_release-notes for details. Failure to read and understand the documentation for this and the implications can result in outages and data loss."
```

Thealisyed replied:
Thanks, have fixed this link

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+	adminGateConfigMap.Data[GatewayAPIAdminKey] = GatewayAPIAdminMsg
+
+	log.Info("Adding admin gate for Gateway API management")
+	if err := r.client.Patch(ctx, adminGateConfigMap, client.MergeFrom(originalAdminGateConfigMap)); err != nil {
+		return fmt.Errorf("failed to update configmap %s: %w", operatorcontroller.AdminGatesConfigMapName(), err)
```
You can do `r.clientUpdate()` instead, like we do in most of our controllers. This will avoid the need for the copying of the admingates configmap. Btw your error message says `update` too.

Thealisyed replied:
Corrected this!

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+
+// The addAdminGate function is responsible for adding the admin gate to the ConfigMap.
+func (r *reconciler) addAdminGate(ctx context.Context) error {
+	adminGateConfigMap := &corev1.ConfigMap{}
```
```suggestion
	adminGatesConfigMap := &corev1.ConfigMap{}
```

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,169 @@
+
+// The removeAdminGate function is responsible for removing the admin gate from the ConfigMap.
+func (r *reconciler) removeAdminGate(ctx context.Context) error {
+	adminGateConfigMap := &corev1.ConfigMap{}
```
```suggestion
	adminGatesConfigMap := &corev1.ConfigMap{}
```

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+		return
+	}
+
+	_, adminGateKeyExists := configMap.Data[gatewayapi_upgradeable.GatewayAPIAdminKey]
```
The tests should avoid using the same variables as used in the code they test. Otherwise we risk to repeat mistakes made in the code.
```suggestion
	_, adminGateKeyExists := configMap.Data["ack-gateway-api-management"]
```

And please unexport the admin key and admin message variables from the controller package.

Thealisyed replied:
Fixed

alebedev87 commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+func createCRDs(t *testing.T) {
+	crds := []*apiextensionsv1.CustomResourceDefinition{
+		manifests.GatewayClassCRD(),
+		manifests.GatewayCRD(),
+		manifests.HTTPRouteCRD(),
+		manifests.ReferenceGrantCRD(),
+	}
```
Same remark about using the code in the tests. Your controller checks an existence of any CRD from 2 API groups. For the purpose of your test it's enough to build 2 test CRDs (see [an example of a test CRD](https://github.com/openshift/cluster-ingress-operator/blob/7dc12cf28557f8c80460435649cfc393761e0285/test/e2e/util_gatewayapi_test.go#L353-L369)).

Thealisyed replied:
Fixed after rebased with Greg's [PR](https://github.com/openshift/cluster-ingress-operator/pull/1193/commits)

alebedev87 commented:
@Thealisyed : can you please update the PR description too?

shaneutt reviewed (commented):
Did another pass. Most of my comments are covered already by existing comments :+1:

Thealisyed commented:
This PR introduces the following:

Initialize the GatewayAPI upgradeable controller

Added the required imports and constants.
Defined predicates for CRD and ConfigMap events to manage the admin gate.
Configured watches in the controller for specific changes in the ConfigMap.
Implement the reconciler logic

Added the Reconcile function to manage the admin gate addition and removal based on conditions
Enhance the operator initialization

Conditionally initialize the gatewayapi_upgradeable controller based on the GatewayAPI feature gate.
End-to-End Tests

Added the TestGatewayAPIUpgradeable end-to-end test.
Included the new test case in the test suite.

alebedev87 reviewed (approved):
Removing "requested changes".

Thealisyed commented:
/retest 

Time out failures and unrelated to changes

Thealisyed commented:
/retest 

Cluster Authentication, timeout and unrelated to changes failure again

Thealisyed commented:
/retest 

TestHeaderName test is failing unrelated to changes

Thealisyed commented:
/jira refresh

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,163 @@
+const (
+	controllerName     = "gatewayapi_upgradeable_controller"
+	gatewayAPIAdminKey = "ack-gateway-api-management"
+	gatewayAPIAdminMsg = "Gateway API CRDs have been detected. OCP fully manages the life-cycle of Gateway API CRDs. External management is unsupported and will be prevented. Tampering with the prevention mechanism will cause the cluster to move to a 'Degraded' state. The cluster administrator is responsible for the safety of existing Gateway API implementations and must acknowledge their responsibi ...
```
I would like to avoid mentioning any details from the development which is currently ongoing (https://issues.redhat.com/browse/NE-1969).
```suggestion
	gatewayAPIAdminMsg = "Gateway API CRDs have been detected. OCP fully manages the life-cycle of Gateway API CRDs. External management is unsupported and will be prevented. The cluster administrator is responsible for the safety of existing Gateway API implementations and must acknowledge their responsibilities via the admin gate to proceed with upgrades. See https://docs.redhat.com/en/documentation/openshift_container_platform/4.19/html/release_notes/ocp-4-19-release-notes#ocp-4-19-networking-gateway-api-crd-lifecycle_release-notes for details. Failure to read and understand the documentation for this and the implications can result in outages and data loss."
```

Thealisyed replied:
Makes sense, fixed it!

alebedev87 commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,163 @@
+
+const (
+	controllerName     = "gatewayapi_upgradeable_controller"
+	gatewayAPIAdminKey = "ack-gateway-api-management"
```
> The format of the admin gate should mention the version it applies to.

And maybe the version it blocks the upgrade to as well (e.g. `ack-4.18-gateway-api-management-in-4.19`). Something to be checked with @wking.

alebedev87 replied:
For the reference, th EP which describes the format: https://github.com/openshift/enhancements/blob/16ec2e8ba16e8cd53e3744b9678aa02c4f3f4f32/enhancements/update/upgrades-blocking-on-ack.md.

Thealisyed replied:
Had a look at EP and yes you are correct. I fixed the formate of the admin gate

Miciah commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,163 @@
+const (
+	controllerName     = "gatewayapi_upgradeable_controller"
+	gatewayAPIAdminKey = "ack-gateway-api-management"
+	gatewayAPIAdminMsg = "Gateway API CRDs have been detected. OCP fully manages the life-cycle of Gateway API CRDs. External management is unsupported and will be prevented. The cluster administrator is responsible for the safety of existing Gateway API implementations and must acknowledge their responsibilities via the admin gate to proceed with upgrades. See https://docs.redhat.com/en/documentati ...
```
https://github.com/openshift/openshift-docs/pull/89357 adds the 4.19 release notes outline, with a placeholder section for this link to target.  We'll need to check with the docs team as to when the docs will go live.  

(Even once it initially goes "live", that just means it'll say "We will have details here when OpenShift Container Platform 4.19 is released" instead of a 404.)

wking replied:
So do y'all want to create a KCS or other document you can release independently, so folks who see this message while running a 4.18 release can onboard themselves with that context without waiting for 4.19 to GA?  They could theoretically be working through walking back any Gateway API CRD managers they run today, if they had access to those docs, and be ready to move to 4.19 on GA-day.

Miciah commented on pkg/operator/controller/gatewayapi-upgradeable/controller.go:
```diff
@@ -0,0 +1,163 @@
+	return c, nil
+}
+
+// Reconciler struct holds the client and cache attributes.
```
Make sure the name in the godoc matches the type's actual name.  
```suggestion
// reconciler struct holds the client and cache attributes.
```

Miciah commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,90 @@
+	} else if gatewayAPIEnabled {
+		t.Skip("Gateway API is enabled, skipping TestGatewayAPIUpgradeable")
+	}
+	createCRDs(t)
+	testAdminGate(t, true)
+	deleteExistingCRD(t, "gatewayclasses.gateway.networking.k8s.io")
```
The test helpers are inconsistent:

- `createCRDs` appears to create only 1 CRD, so the name is misleading.
- The definition of `createCRDs` hard-codes a CRD whereas `deleteExistingCRD` takes the CRD name as a parameter, which is inconsistent.  

Aside from that, the test should have a `t.Cleanup` handler to delete the CRD that `createCRDs` creates.  Suppose you ran the test twice and that `testAdminGate` failed the first time; then the second time you ran the test, `createCRDs` would fail because the test would have left behind the CRD the first time around.  Where feasible, tests should be idempotent.  

It would also be helpful to have godoc on `TestGatewayAPIUpgradeable` to explain what it does.

Thealisyed replied:
After rebasing [PR](https://github.com/openshift/cluster-ingress-operator/pull/1193), `createCRDs` is reflective as it creates a range of CRDs. 
`t.Cleanup` was [added](https://github.com/openshift/cluster-ingress-operator/pull/1196/files#diff-28e96b945950136ab0fee2af650b3c6b166d1ae94ea2ebc12b42269dbb054bb8R36:~:text=t.Cleanup,%7D)) to check if any CRDs are left over and removes them.
A [godoc](https://github.com/openshift/cluster-ingress-operator/pull/1196/files#diff-28e96b945950136ab0fee2af650b3c6b166d1ae94ea2ebc12b42269dbb054bb8R36:~:text=//%20TestGatewayAPIUpgradeable%20verifies%20the,gate%20is%20disabled.) was added for `TestGatewayAPIUpgrdeable`

Miciah commented:
This is a low-risk change.  The impact is limited to a single new controller, similar to the one in https://github.com/openshift/cluster-ingress-operator/pull/1014/commits/6778ce1a619b2f58310a037694e341e446888c55, to update the "admin-gates" configmap.  The logic is well contained and has E2E test coverage.  

/label backport-risk-assessed

This is a low-risk change. The impact is limited to the operator's clusteroperator status reporting logic, it is well contained, and it has good test coverage. (Thanks for the detailed comments and commit messages!)

alebedev87 commented:
@Thealisyed: Thanks for addressing all the comments!

/lgtm

alebedev87 commented:
/retest

grzpiotrowski commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -80,3 +96,22 @@ func deleteExistingCRDs(t *testing.T, crds []*apiextensionsv1.CustomResourceDefi
+		t.Fatalf("Timed out trying to get admin-gates configmap: %v", err)
+	}
+
+	_, adminGateKeyExists := adminGatesConfigMap.Data["ack-gateway-api-management"]
```
I saw some failed tests and got curious as it seemed to be fine earlier.
Is this the culprit and should be `ack-4.18-gateway-api-management-in-4.19`?

alebedev87 replied:
> Is this the culprit and should be ack-4.18-gateway-api-management-in-4.19?

Yes. Good catch!

Thealisyed replied:
Added that! However I can see the test still failing for some reason

Thealisyed replied:
```
=== NAME  TestAll/parallel/TestGatewayAPIUpgradeable
    gatewayapi_upgradeable_test.go:58: Expected admin gate key existence to be true, but got false
```

e2e-aws-operator

Thealisyed replied:
Fixed that, thanks for that catch @grzpiotrowski!

alebedev87 commented:
/lgtm

candita commented:
@Thealisyed I see many comments resolved with no answer from you. Can you please take a quick look through and add responses that explain what you did to address the comment?  Even just a "done", or thumbs-up would suffice.  Our team generally does not resolve an un-answered comment.

candita commented:
/test e2e-aws-operator

candita commented:
> == NAME  TestAll/parallel/TestAWSEIPAllocationsForNLB
>     util_test.go:1068: service openshift-ingress/openshift-ingress UID has not changed yet, retrying...
>     lb_eip_test.go:149: failed to delete and recreate service: context deadline exceeded

/test e2e-aws-operator-techpreview

alebedev87 reviewed (approved):
Removing "change requested" [(change)](https://github.com/openshift/cluster-ingress-operator/compare/bd93c36050048bd025308a30d08d8373723b71ca..c22c21d56ad6f0169d72f3890d6c17ab58c7a44f).

Thealisyed commented:
/retest 

Static nodes not created and other unrelated to changes

alebedev87 commented:
Build of the CIO image took too much time:
```
Build cluster-ingress-operator-amd64 succeeded after 1h4m8s 
```

Due to some problems with the internal image registry:
```
Pushing image image-registry.openshift-image-registry.svc:5000/ci-op-0s543lfi/pipeline:cluster-ingress-operator-amd64 ...
Warning: Push failed, retrying in 5s ...
Getting image source signatures
Warning: Push failed, retrying in 5s ...
Getting image source signatures
Warning: Push failed, retrying in 5s ...
Registry server Address: 
Registry server User Name: <token>
Registry server Email: 
Registry server Password: <<non-empty>>
error: build error: Failed to push image: trying to reuse blob sha256:25c75c34b2e2b68ba9245d9cddeb6b8a0887371ed30744064f85241a75704d87 at destination: unable to retrieve auth token: invalid username/password: authentication required
```

The CI job was forcibly stopped because the default 4h timeout was reached.

/retest-required

Thealisyed commented:
@candita Thanks for the feedback, and my apologies for missing that. I’ve been resolving the comments without adding responses to the reviewer, which is something Miciah and I had discussed earlier. I’ll go through the PR now and make sure to address the remaining comments.

Thealisyed commented:
/test e2e-aws-ovn-serial

Due to some problems with the internal image registry:


```
error: build error: Failed to push image: trying to reuse blob sha256:25c75c34b2e2b68ba9245d9cddeb6b8a0887371ed30744064f85241a75704d87 at destination: unable to retrieve auth token: invalid username/password: authentication required

{"component":"entrypoint","file":"sigs.k8s.io/prow/pkg/entrypoint/run.go:169","func":"sigs.k8s.io/prow/pkg/entrypoint.Options.ExecuteProcess","level":"error","msg":"Process did not finish before 4h0m0s timeout","severity":"error","time":"2025-03-21T04:34:06Z"}

```
The CI job was forcibly stopped because the default 4h timeout was reached.
Nothing to suggest issues with test so will retest now

Thealisyed commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+
+func testAdminGate(t *testing.T, shouldExist bool) {
+	t.Helper()
+	//	crdsExist := checkCRDsExist(t)
```
Thanks, removed now

Thealisyed commented on test/e2e/gatewayapi_upgradeable_test.go:
```diff
@@ -0,0 +1,89 @@
+	testAdminGate(t, false)
+}
+
+func createCRDs(t *testing.T) {
```
Added helper

Thealisyed commented:
/test e2e-aws-ovn-serial

auth error: failed to push image

`error: build error: Failed to push image: trying to reuse blob sha256:25c75c34b2e2b68ba9245d9cddeb6b8a0887371ed30744064f85241a75704d87 at destination: unable to retrieve auth token: invalid username/password: authentication required
`

Failed at 4 hour timeout:

`Process did not finish before 4h0m0s timeout`

It has failed before in Job history for this reason so retesting

alebedev87 commented:
/retest-required
